### PR TITLE
Support exclusive lists with GoToSocial 0.17

### LINKED ISSUE #808: Support exclusive lists with GotoSocial 0.17
**Is your feature request related to a problem? Please describe.**

GotoSocial's upcoming 0.17 release will include Mastodon-API-compatible exclusive lists. Phanpy already supports them on Mastodon instances through the "Hide posts on this list from Home/Following" list edit checkbox, and they work great.

**Describe the solution you'd like**

I'd love to see Phanpy add support for them with GtS 0.17 as well. API should be the same:
- superseriousbusiness/gotosocial#3280

Feditext and Fedicat didn't hit any issues adding it, afaik.

### PR DESCRIPTION
**Pull request by graue**
Fixes #808: supports exclusive lists with GoToSocial 0.17+.

Unlike some other server types, GTS does not use version strings that pretend to be Mastodon (for example, Pixelfed's version string of "3.5.3 (compatible; Pixelfed 0.12.3)". Instead, to detect that the server is GTS, we must use [NodeInfo](https://github.com/jhass/nodeinfo). I've added support for this ~~and rewritten the feature detection to use it~~.

Note that for Mastodon <4.3 (all current stable releases of Mastodon), the NodeInfo request will fail due to mastodon/mastodon#23135, so we'll fall back to the existing behavior. Mastodon 4.3 fixes this. For other server software, this will allow for more accurate checking of feature availability.

### CODE REVIEW

**Comment by cheeaun** [comment]
Honestly, this changes too many things at once, which makes it hard for me to review.

Few notes:
- Could be simplified by just introducing `@gotosocial/list-exclusive`.
- `@mastodon/list-exclusive': [['mastodon', '>=4.2'], ['gotosocial', '>=0.17']],` looks confusing to me, thus the point above.
- `versionSatisfies(software, version, [softwareSpec, versionSpec])` is kinda unreadable to me
- I'll still prefer that current implementation is untouched and add nodeinfo check **only** for platforms that don't add their own user agent with Mastodon's, for now. Try not to over-engineer this as there could be many more platforms in the future that might do different things, especially learning from the history of browser user agent sniffing 🙈

**Comment by graue** [comment]
Thanks, yeah, that part could definitely be written better. I can take a second pass on readability and split it into multiple commits, although it also sounds like you don't like this approach at all...

> Could be simplified by just introducing `@gotosocial/list-exclusive`.

I thought you might prefer only to have to check one flag at the point where the feature is implemented. This could one day also become `@akkoma/list-exclusive`, `@pixelfed/list-exclusive`...

> I'll still prefer that current implementation is untouched and add nodeinfo check only for platforms that don't add their own user agent with Mastodon's, for now.

The issue here is, without getting the nodeinfo, Phanpy doesn't know it's a platform that doesn't add its own user agent with Mastodon's. Unless the code did something like assume that a version >=3 is Mastodon and <3 is GTS, but that doesn't seem very future-proof.

Nodeinfo has become a widely used standard, with this [list of implementations](https://github.com/jhass/nodeinfo?tab=readme-ov-file#support) including every kind of server Phanpy has issues about (except one, you have an issue tagged Takahe which isn't in the list but I checked and it [also supports nodeinfo](https://takahe.mercadosocial.madrid/)).

**Comment by graue** [comment]
All right, here's a smaller version of the change. It feels a bit less future-proof to me than what I was trying to do, but I admit my implementation was awkward. Perhaps if projects start using the NodeInfo more heavily to communicate their capabilities, this could be revisited later.

**Comment by graue** [comment]
@cheeaun, any thoughts on my revised version of this based on your feedback?

For what it's worth (side note...), with Mastodon 4.3 out now, the nodeInfo API will become almost universally supported as servers upgrade.

File: src/utils/api.js [hunk 1]
```diff
@@ -89,6 +88,29 @@ export async function initInstance(client, instance) {
...
+      const wellKnown = await (await fetch(`${urlBase}/.well-known/nodeinfo`)).json();
+      if (Array.isArray(wellKnown?.links)) {
+        const nodeInfoUrl = wellKnown.links.find(
+          (link) => typeof link.rel === 'string' &&
+          link.rel.startsWith('http://nodeinfo.diaspora.software/ns/schema/')
+        )?.href;
+        if (nodeInfoUrl && nodeInfoUrl.startsWith(urlBase)) {
+          const nodeInfo = await (await fetch(nodeInfoUrl)).json();
+          if (typeof nodeInfo?.software?.name === 'string') {
+            info.software_name = nodeInfo.software.name;
```
Comment (cheeaun) [inline src/utils/api.js:106]: Re-reading this, it seems useful to store the whole nodeinfo object somewhere else (maybe on 'nodeinfo'/'nodeinfos' key on localStorage, separate from 'instances') and not add an extra key to "info" object here to make sure there's no potential future clash.

File: src/utils/supports.js [hunk 2]
```diff
@@ -25,11 +26,13 @@ const platformFeatures = {
   '@pleroma/local-visibility-post': containPleroma,
   '@akkoma/local-visibility-post': containAkkoma,
 };
+
 const supportsCache = {};
 
 function supports(feature) {
   try {
-    const { version, domain } = getCurrentInstance();
+    let { version, domain, software_name } = getCurrentInstance();
```
Comment (cheeaun) [inline src/utils/supports.js:34]: If nodeinfo object is stored separately, possibly need another `getCurrentNodeInfo` function. So this could look like this:

```
const { software: { name: softwareName } } = getCurrentNodeInfo();
```

File: src/utils/supports.js [hunk 3]
```diff
@@ -39,10 +42,13 @@ function supports(feature) {
 
     const range = features[feature];
     if (!range) return false;
-    return (supportsCache[key] = satisfies(version, range, {
-      includePrerelease: true,
-      loose: true,
-    }));
+    return (supportsCache[key] = (
+      containGTS.test(feature) === containGTS.test(software_name)
```
Comment (cheeaun) [inline src/utils/supports.js:46]: This line is honestly a bit hard to read. I get it, if it's Mastodon, it'll be `false === false` which returns `true` 😬

Could possibly prepend this above instead? e.g.:

```
if (softwareName.toLowerCase() === 'gotosocial' && feature.startsWith('@gotosocial') {
  return satisfies(version, range, {
    includePrerelease: true,
    loose: true,
  });
}
```

**Review by cheeaun — CHANGES_REQUESTED** [review]

File: src/utils/supports.js (hunk 3 above)
Comment (graue) [reply src/utils/supports.js:46]: Hmm, that doesn't quite work, because if the feature is GoToSocial but the server is Mastodon, or vice versa, and if it then falls through to the original `satisfies` check, it's going to give incorrect results. We need some notion of "Does the software that's running on this server match the software this feature applies to?"

**Comment by graue** [comment]
Rebased and addressed feedback. Hopefully this will do the trick for you!

File: src/utils/supports.js [hunk 4]
```diff
@@ -39,10 +48,17 @@ function supports(feature) {
 
     const range = features[feature];
     if (!range) return false;
-    return (supportsCache[key] = satisfies(version, range, {
-      includePrerelease: true,
-      loose: true,
-    }));
+
+    // '@mastodon/blah' => 'mastodon'
+    const featureSoftware = feature.match(/^@([a-z]+)\//)[1];
```
Comment (cheeaun) [inline src/utils/supports.js:53]: This needs to be like `(feature.match(/^@([a-z]+)\//) || [,false])[1]` because if there's no match, `match` will return `null`.

I know my code above looks fancy (it's in other files) so feel free to write it in a more understandable way.

File: src/utils/supports.js [hunk 5]
```diff
@@ -25,11 +26,19 @@ const platformFeatures = {
...
 const supportsCache = {};
 
 function supports(feature) {
   try {
-    const { version, domain } = getCurrentInstance();
+    let { version, domain } = getCurrentInstance();
+    let softwareName = getCurrentNodeInfo()?.software?.name || 'mastodon';
+
+    if (softwareName === 'hometown') {
+      // Hometown is a Mastodon fork and inherits its features
```
Comment (cheeaun) [inline src/utils/supports.js:38]: Wondering, how is this relevant to the current PR, which is meant for GTS?

Comment (graue) [reply src/utils/supports.js:38]: When a future Hometown release includes the fix for nodeInfo, `softwareName` will be `'hometown'`, so for features like `'@mastodon/list-exclusive'`, this comparison (`'mastodon' === 'hometown'`) will be false:

```.js
const doesSoftwareMatch = featureSoftware === softwareName.toLowerCase();
```

File: src/utils/supports.js (hunk 4 above)
Comment (graue) [reply src/utils/supports.js:53]: I thought I could rely on `supports()` only being called with strings that match the `@software/feature-name` pattern; currently, this is the case:

```
$ git grep supports\(
src/components/account-info.jsx:                {supports('@mastodon/profile-private-note') && (
src/components/account-info.jsx:              supports('@mastodon/profile-edit') && (
src/components/compose.jsx:                if (editStatus && supports('@mastodon/edit-media-attributes')) {
src/components/compose.jsx:                {(supports('@pleroma/local-visibility-post') ||
src/components/compose.jsx:                  supports('@akkoma/local-visibility-post')) && (
src/components/compose.jsx:  const supportsEdit = supports('@mastodon/edit-media-attributes');
... etc., all of the calls are like this
```

So it seems like a call which did not match that pattern would likely be a bug. Do you want me to check for that here anyway?

**Comment by graue** [comment]
When you get a chance, can you advise me on if you still think this needs changing, and my questions above? Thanks!

**Comment by cheeaun** [comment]
@graue sorry, progress has been slow on my side. Looks good so far, is it possible to resolve the conflicts?

**Comment by graue** [comment]
I'm a bit confused about how the `.po` files are generated or what they're supposed to contain. How do you recommend fixing the conflict? Can I just regenerate en.po from the source?

**Review by cheeaun — APPROVED** [review]
